substratevm/src/com.oracle.svm.reflect/src/com/oracle/svm/reflect/serialize/hosted/SerializationFeature.java

```diff
@@ -205,10 +223,69 @@ private void abortIfSealed() {
         UserError.guarantee(!sealed, "Too late to add classes for serialization. Registration must happen in a Feature before the analysis has finished.");
     }
 
+    @Override
+    public void registerAllAssociatedClasses(ConfigurationCondition condition, Class<?> clazz) {
+        String targetClassName = clazz.getName();
+        // If the serialization target is primitive, it needs to get boxed, because the target is
+        // always an Object.
+        if (clazz.isPrimitive()) {
+            Class<?> boxedType = JavaKind.fromJavaClass(clazz).toBoxedJavaClass();
+            registerAllAssociatedClasses(condition, boxedType);
+            return;
+        } else if (!Serializable.class.isAssignableFrom(clazz)) {
+            println("Warning: The given class " + targetClassName + " is not serializable, it wouldn't get registered.\n");
+            return;
+        } else if (access.findSubclasses(clazz).size() > 1) {
+            println("Warning: The given class " + targetClassName +
+                            " has subclasses. We can't figure out which one should be registered. Skip.\n");
+            return;
+        }
+        try {
+            clazz.getDeclaredMethod("writeObject", ObjectOutputStream.class);
+            println("Warning: The given class " + targetClassName +
+                            " has customized serialization method writeObject. " +
+                            "We can't figure out what should be registered. Skip.\n");
+            return;
+        } catch (NoSuchMethodException e) {
+            // Expected case. Do nothing
+        }
+        register(condition, clazz);
```

Comment:
Provide a private `registerAllAssociatedClasses` that takes a `Set<Class<?>>` parameter and make the `public void registerAllAssociatedClasses` call it with `registerAllAssociatedClasses(condition, clazz, new HashSet<>())`.
In the private `registerAllAssociatedClasses` use
```java
    private void registerAllAssociatedClasses0(ConfigurationCondition condition, Class<?> clazz, Set<Class<?>> alreadyVisited) {
        if (alreadyVisited.contains(clazz)) {
            return;
        }
        alreadyVisited.add(clazz);
        ... rest of impl.
    }
```

Reply:
We already have a set `SerializationSupport.registeredTargets` to track all registered classes. It is checked in `register` method, so that all serialization target class registrations are checked, not only the ones through `registerAllAssociatedClasses`.

Reply:
Yes ... but
you are not using that to prevent redundant reentering of `registerAllAssociatedClasses` for the same class object.

Also note that it's a different thing knowing that you already have a class that was registered with `register` (i.e. having it in `SerializationSupport.registeredTargets` compared to knowing that you already ran `registerAllAssociatedClasses` for a given class. If you would reuse `SerializationSupport.registeredTargets`  for the above it can happen that previously you registered a class with `register` and
```java
if (isRegistered(clazz)) {
    return;
}
```
would prevent running `registerAllAssociatedClasses` although that hasn't happen before. So please implement this as I suggested in https://github.com/oracle/graal/pull/3760#discussion_r770612854

Reply:
I see your point. Fixed as suggested.